### ISSUE TITLE
Cannot connect from Debian 9 client to functioning Debian 8 server

### ISSUE THREAD

**Issue by tpdownes** [globus/gsi-openssh#19]
We've compiled your 7.5 pre-release on Debian 9 stretch (using libssl 1.1 because the globus libraries have compiled that way). We find that we cannot connect to hosts that run Debian 8 and work from Debian 8 and RHEL7 clients. We haven't yet tested in any other combination.

I wonder if related to [this issue in globus toolkit](https://github.com/globus/globus-toolkit/issues/92)?

The client:
```
albert@69800e924f9c:~$ grid-proxy-info
subject  : /DC=org/DC=cilogon/C=US/O=LIGO/CN=Thomas Downes thomas.downes@ligo.org
issuer   : /DC=org/DC=cilogon/C=US/O=CILogon/CN=CILogon Basic CA 1
identity : /DC=org/DC=cilogon/C=US/O=LIGO/CN=Thomas Downes thomas.downes@ligo.org
type     : end entity credential
strength : 2048 bits
path     : /tmp/x509up_u1000
timeleft : 430:15:06  (17.9 days)
albert@69800e924f9c:~$ gsissh -p 443 -vvv tpdownes@pcdev3.ligo.uwm.edu
OpenSSH_7.5p1-GSI -hpn14v13 GSI, OpenSSL 1.1.0f  25 May 2017
debug1: Reading configuration data /etc/gsissh/ssh_config
debug2: resolving "pcdev3.ligo.uwm.edu" port 443
debug2: ssh_connect_direct: needpriv 0
debug1: Connecting to pcdev3.ligo.uwm.edu [129.89.61.93] port 443.
debug1: Connection established.
debug1: key_load_public: No such file or directory
debug1: identity file /container/albert/.ssh/id_rsa type -1
debug1: key_load_public: No such file or directory
debug1: identity file /container/albert/.ssh/id_rsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file /container/albert/.ssh/id_dsa type -1
debug1: key_load_public: No such file or directory
debug1: identity file /container/albert/.ssh/id_dsa-cert type -1
debug1: identity file /container/albert/.ssh/id_ecdsa type 3
debug1: key_load_public: No such file or directory
debug1: identity file /container/albert/.ssh/id_ecdsa-cert type -1
debug1: key_load_public: No such file or directory
debug1: identity file /container/albert/.ssh/id_ed25519 type -1
debug1: key_load_public: No such file or directory
debug1: identity file /container/albert/.ssh/id_ed25519-cert type -1
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_7.5p1-GSI -hpn14v13 GSI
debug1: Remote protocol version 2.0, remote software version OpenSSH_7.1p2-hpn14v10 GSI_GSSAPI_20160606 GSI
debug1: match: OpenSSH_7.1p2-hpn14v10 GSI_GSSAPI_20160606 GSI pat OpenSSH* compat 0x04000000
debug2: fd 3 setting O_NONBLOCK
debug1: Authenticating to pcdev3.ligo.uwm.edu:443 as 'tpdownes'
debug3: Trying to reverse map address 129.89.61.93.
debug1: Offering GSSAPI proposal: gss-group16-sha512-vz8J1E9PzLr8b1K+0remTg==,gss-group18-sha512-vz8J1E9PzLr8b1K+0remTg==,gss-gex-sha1-vz8J1E9PzLr8b1K+0remTg==,gss-group14-sha256-vz8J1E9PzLr8b1K+0remTg==,gss-group14-sha1-vz8J1E9PzLr8b1K+0remTg==,gss-group1-sha1-vz8J1E9PzLr8b1K+0remTg==
debug3: put_host_port: [pcdev3.ligo.uwm.edu]:443
debug3: hostkeys_foreach: reading file "/container/albert/.ssh/known_hosts"
debug3: send packet: type 20
debug1: SSH2_MSG_KEXINIT sent
debug3: receive packet: type 20
debug1: SSH2_MSG_KEXINIT received
debug1: AUTH STATE IS 0
debug2: local client KEXINIT proposal
debug2: KEX algorithms: gss-group16-sha512-vz8J1E9PzLr8b1K+0remTg==,gss-group18-sha512-vz8J1E9PzLr8b1K+0remTg==,gss-gex-sha1-vz8J1E9PzLr8b1K+0remTg==,gss-group14-sha256-vz8J1E9PzLr8b1K+0remTg==,gss-group14-sha1-vz8J1E9PzLr8b1K+0remTg==,gss-group1-sha1-vz8J1E9PzLr8b1K+0remTg==,curve25519-sha256,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group16-sha512,diffie-hellman-group18-sha512,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha256,diffie-hellman-group14-sha1,ext-info-c
debug2: host key algorithms: ecdsa-sha2-nistp256-cert-v01@openssh.com,ecdsa-sha2-nistp384-cert-v01@openssh.com,ecdsa-sha2-nistp521-cert-v01@openssh.com,ssh-ed25519-cert-v01@openssh.com,ssh-rsa-cert-v01@openssh.com,ecdsa-sha2-nistp256,ecdsa-sha2-nistp384,ecdsa-sha2-nistp521,ssh-ed25519,rsa-sha2-512,rsa-sha2-256,ssh-rsa,null
debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,aes192-cbc,aes256-cbc
debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,aes192-cbc,aes256-cbc
debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: compression ctos: none,zlib@openssh.com,zlib
debug2: compression stoc: none,zlib@openssh.com,zlib
debug2: languages ctos:
debug2: languages stoc:
debug2: first_kex_follows 0
debug2: reserved 0
debug2: peer server KEXINIT proposal
debug2: KEX algorithms: gss-gex-sha1-dZuIebMjgUqaxvbF7hDbAw==,gss-group1-sha1-dZuIebMjgUqaxvbF7hDbAw==,gss-group14-sha1-dZuIebMjgUqaxvbF7hDbAw==,curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1
debug2: host key algorithms: ssh-rsa,ecdsa-sha2-nistp256,ssh-ed25519
debug2: ciphers ctos: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: ciphers stoc: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
debug2: MACs ctos: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: MACs stoc: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
debug2: compression ctos: none,zlib@openssh.com
debug2: compression stoc: none,zlib@openssh.com
debug2: languages ctos:
debug2: languages stoc:
debug2: first_kex_follows 0
debug2: reserved 0
debug1: kex: algorithm: curve25519-sha256@libssh.org
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: REQUESTED ENC.NAME is 'chacha20-poly1305@openssh.com'
debug1: kex: server->client cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug1: REQUESTED ENC.NAME is 'chacha20-poly1305@openssh.com'
debug1: kex: client->server cipher: chacha20-poly1305@openssh.com MAC: <implicit> compression: none
debug3: send packet: type 30
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug3: receive packet: type 31
debug1: Server host key: ecdsa-sha2-nistp256 SHA256:J0P/JNad2NRXvYeztdcheXzxdvW7rF3oVT+dvTieAeY
debug3: put_host_port: [129.89.61.93]:443
debug3: put_host_port: [pcdev3.ligo.uwm.edu]:443
debug3: hostkeys_foreach: reading file "/container/albert/.ssh/known_hosts"
debug3: hostkeys_foreach: reading file "/container/albert/.ssh/known_hosts"
debug1: checking without port identifier
debug3: hostkeys_foreach: reading file "/container/albert/.ssh/known_hosts"
debug3: record_hostkey: found key type ECDSA in file /container/albert/.ssh/known_hosts:1
debug3: load_hostkeys: loaded 1 keys from pcdev3.ligo.uwm.edu
debug3: hostkeys_foreach: reading file "/container/albert/.ssh/known_hosts"
debug3: record_hostkey: found key type ECDSA in file /container/albert/.ssh/known_hosts:1
debug3: load_hostkeys: loaded 1 keys from 129.89.61.93
debug1: Host 'pcdev3.ligo.uwm.edu' is known and matches the ECDSA host key.
debug1: Found key in /container/albert/.ssh/known_hosts:1
debug1: found matching key w/out port
debug3: send packet: type 21
debug2: set_newkeys: mode 1
debug1: rekey after 134217728 blocks
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug3: receive packet: type 21
debug1: SSH2_MSG_NEWKEYS received
debug2: set_newkeys: mode 0
debug1: rekey after 134217728 blocks
debug2: key: /container/albert/.ssh/id_rsa ((nil))
debug2: key: /container/albert/.ssh/id_dsa ((nil))
debug2: key: /container/albert/.ssh/id_ecdsa (0x55838c68a730)
debug2: key: /container/albert/.ssh/id_ed25519 ((nil))
debug3: send packet: type 5
debug3: receive packet: type 6
debug2: service_accept: ssh-userauth
debug1: SSH2_MSG_SERVICE_ACCEPT received
debug3: send packet: type 50
debug3: receive packet: type 51
debug1: Authentications that can continue: publickey,gssapi-keyex,gssapi-with-mic
debug3: start over, passed a different list publickey,gssapi-keyex,gssapi-with-mic
debug3: preferred gssapi-keyex,gssapi-with-mic,publickey,keyboard-interactive,password
debug3: authmethod_lookup gssapi-keyex
debug3: remaining preferred: gssapi-with-mic,publickey,keyboard-interactive,password
debug3: authmethod_is_enabled gssapi-keyex
debug1: Next authentication method: gssapi-keyex
debug1: No valid Key exchange context
debug2: we did not send a packet, disable method
debug3: authmethod_lookup gssapi-with-mic
debug3: remaining preferred: publickey,keyboard-interactive,password
debug3: authmethod_is_enabled gssapi-with-mic
debug1: Next authentication method: gssapi-with-mic
debug3: send packet: type 50
debug2: we sent a gssapi-with-mic packet, wait for reply
debug3: receive packet: type 51
debug1: Authentications that can continue: publickey,gssapi-keyex,gssapi-with-mic
debug2: we did not send a packet, disable method
debug3: authmethod_lookup publickey
debug3: remaining preferred: keyboard-interactive,password
debug3: authmethod_is_enabled publickey
debug1: Next authentication method: publickey
debug1: Trying private key: /container/albert/.ssh/id_rsa
debug3: no such identity: /container/albert/.ssh/id_rsa: No such file or directory
debug1: Trying private key: /container/albert/.ssh/id_dsa
debug3: no such identity: /container/albert/.ssh/id_dsa: No such file or directory
debug1: Offering ECDSA public key: /container/albert/.ssh/id_ecdsa
debug3: send_pubkey_test
debug3: send packet: type 50
debug2: we sent a publickey packet, wait for reply
debug3: receive packet: type 51
debug1: Authentications that can continue: publickey,gssapi-keyex,gssapi-with-mic
debug1: Trying private key: /container/albert/.ssh/id_ed25519
debug3: no such identity: /container/albert/.ssh/id_ed25519: No such file or directory
debug2: we did not send a packet, disable method
debug1: No more authentication methods to try.
Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
```
The server (443 because it's firewall friendly without being the active ssh port)
```
root@pcdev3:~# /usr/sbin/gsisshd -p 443 -d
debug1: Config token is acceptenv
debug1: Config token is challengeresponseauthentication
debug1: Config token is passwordauthentication
debug1: Config token is permitrootlogin
debug1: Config token is pidfile
debug1: Config token is printmotd
debug1: Config token is protocol
debug1: Config token is pubkeyauthentication
debug1: Config token is subsystem
debug1: Config token is usepam
debug1: Config token is useprivilegeseparation
debug1: Config token is x11forwarding
debug1: Config token is match
debug1: Config token is x11forwarding
debug1: Config token is allowtcpforwarding
debug1: Config token is allowagentforwarding
debug1: Config token is gssapidelegatecredentials
debug1: HPN Buffer Size: 65536
debug1: sshd version OpenSSH_7.1, OpenSSL 1.0.1t  3 May 2016
debug1: private host key #0: ssh-rsa SHA256:vRCqoBEYCY79U86ZL+403X7bz9K+wVjyIAvc3wVk9Ws
debug1: private host key #1: ssh-dss SHA256:4oN9gltqNVX2l4Daikcq4G5DWGffsbdd9rIQAhg1bPQ
debug1: private host key #2: ecdsa-sha2-nistp256 SHA256:J0P/JNad2NRXvYeztdcheXzxdvW7rF3oVT+dvTieAeY
debug1: private host key #3: ssh-ed25519 SHA256:jbJzwuRwR9g7GsA7+krym9zU5BmXBCzpeeAJ4OJPdDg
debug1: rexec_argv[0]='/usr/sbin/gsisshd'
debug1: rexec_argv[1]='-p'
debug1: rexec_argv[2]='443'
debug1: rexec_argv[3]='-d'
debug1: Set /proc/self/oom_score_adj from 0 to -1000
debug1: Bind to port 443 on 0.0.0.0.
debug1: Server TCP RWIN socket size: 65536
debug1: HPN Buffer Size: 65536
Server listening on 0.0.0.0 port 443.
debug1: Bind to port 443 on ::.
debug1: Server TCP RWIN socket size: 65536
debug1: HPN Buffer Size: 65536
Server listening on :: port 443.
debug1: Server will not fork when running in debugging mode.
debug1: rexec start in 5 out 5 newsock 5 pipe -1 sock 8
debug1: inetd sockets after dupping: 3, 3
Connection from 129.89.57.180 port 55701 on 129.89.61.93 port 443
debug1: HPN Disabled: 0, HPN Buffer Size: 65536
debug1: Client protocol version 2.0; client software version OpenSSH_7.5p1-GSI -hpn14v13 GSI
SSH: Server;Ltype: Version;Remote: 129.89.57.180-55701;Protocol: 2.0;Client: OpenSSH_7.5p1-GSI -hpn14v13 GSI
debug1: match: OpenSSH_7.5p1-GSI -hpn14v13 GSI pat OpenSSH* compat 0x04000000
debug1: Enabling compatibility mode for protocol 2.0
debug1: Local version string SSH-2.0-OpenSSH_7.1p2-hpn14v10 GSI_GSSAPI_20160606 GSI
debug1: list_hostkey_types: ssh-rsa,ecdsa-sha2-nistp256,ssh-ed25519
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: AUTH STATE IS 0
debug1: REQUESTED ENC.NAME is 'chacha20-poly1305@openssh.com'
debug1: kex: client->server chacha20-poly1305@openssh.com <implicit> none
SSH: Server;Ltype: Kex;Remote: 129.89.57.180-55701;Enc: chacha20-poly1305@openssh.com;MAC: <implicit>;Comp: none
debug1: REQUESTED ENC.NAME is 'chacha20-poly1305@openssh.com'
debug1: kex: server->client chacha20-poly1305@openssh.com <implicit> none
debug1: expecting SSH2_MSG_KEX_ECDH_INIT
debug1: SSH2_MSG_NEWKEYS sent
debug1: expecting SSH2_MSG_NEWKEYS
debug1: SSH2_MSG_NEWKEYS received
debug1: KEX done
debug1: userauth-request for user tpdownes service ssh-connection method none
SSH: Server;Ltype: Authname;Remote: 129.89.57.180-55701;Name: tpdownes
debug1: attempt 0 failures 0
debug1: Config token is acceptenv
debug1: Config token is challengeresponseauthentication
debug1: Config token is passwordauthentication
debug1: Config token is permitrootlogin
debug1: Config token is pidfile
debug1: Config token is printmotd
debug1: Config token is protocol
debug1: Config token is pubkeyauthentication
debug1: Config token is subsystem
debug1: Config token is usepam
debug1: Config token is useprivilegeseparation
debug1: Config token is x11forwarding
debug1: Config token is match
debug1: user tpdownes does not match group list lvcshared at line 16
debug1: Config token is x11forwarding
debug1: Config token is allowtcpforwarding
debug1: Config token is allowagentforwarding
debug1: Config token is gssapidelegatecredentials
debug1: PAM: initializing for "tpdownes"
debug1: PAM: setting PAM_RHOST to "129.89.57.180"
debug1: PAM: setting PAM_TTY to "ssh"
Failed none for tpdownes from 129.89.57.180 port 55701 ssh2
debug1: userauth-request for user tpdownes service ssh-connection method gssapi-with-mic
debug1: attempt 1 failures 0
Failed gssapi-with-mic for tpdownes from 129.89.57.180 port 55701 ssh2
debug1: userauth-request for user tpdownes service ssh-connection method publickey
debug1: attempt 2 failures 0
debug1: test whether pkalg/pkblob are acceptable
debug1: temporarily_use_uid: 1296/502 (e=0/0)
debug1: trying public key file /home/tpdownes/.ssh/authorized_keys
debug1: fd 5 clearing O_NONBLOCK
debug1: restore_uid: 0/0
debug1: temporarily_use_uid: 1296/502 (e=0/0)
debug1: trying public key file /home/tpdownes/.ssh/authorized_keys2
debug1: Could not open authorized keys '/home/tpdownes/.ssh/authorized_keys2': No such file or directory
debug1: restore_uid: 0/0
Failed publickey for tpdownes from 129.89.57.180 port 55701 ssh2
Connection closed by 129.89.57.180
debug1: do_cleanup
debug1: PAM: cleanup
```

**Comment by tpdownes** [comment]
Builds [here](http://software.ligo.org/gridtools/debian/dists/stretch-proposed/).

**Comment by ysvenkat** [comment]
Thanks for the test and feedback. As expected, the client is using the new MIC OID (vz8J1E9PzLr8b1K+0remTg==)  in the KEX proposals, as in: gss-group16-sha512-vz8J1E9PzLr8b1K+0remTg==
but the server is too old to offer KEX algorithms using the new MIC OID. Support for the new MIC OID was introduced in 7.1p2g, with fixes in 7.1p2j.

https://github.com/globus/gsi-openssh/releases/tag/7.1p2g
https://github.com/globus/gsi-openssh/releases/tag/7.1p2j

If you can try upgrading the server to the latest 7.1pj or 7.3p1c release, it should work. Thanks.

**Comment by tpdownes** [comment]
We'll try that. I did not find that setting `GSISSH_FORCE_CONF_FLAG` in the client had any effect. Should it have?

**Comment by ysvenkat** [comment]
No, it wouldn't have made a difference since, from the debug log, we see that the issue is the Debian 8 and CentOS servers not running the latest 7.p2j or 7.3p1c releases of GSI-OpenSSH which are compatible with gsisshd using OpenSSL 1.1 (Debian 9).

**Comment by tpdownes** [comment]
Ah, now I see what you are saying the in comment. I can report the following:

- I upgraded a wheezy machine to `gsi-openssh-server=7.3p1c-1+gt6.wheezy`. I chose the machine because it already had a host certificate but no one (besides me) would care if certificate authentication didn't work.
- I connected to that machine from a docker container running `gsi-openssh-clients=7.5p1-0.1+gt6.0deb9u0`

Upping the verbosity of the client confirms the connection is through GSI. Problem solved. Thank you.
```
debug1: Authentication succeeded (gssapi-keyex).
```
